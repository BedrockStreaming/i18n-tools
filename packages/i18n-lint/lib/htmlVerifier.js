'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _isHtml = require('is-html');

var _isHtml2 = _interopRequireDefault(_isHtml);

var _utils = require('./utils');

var _logger = require('./logger');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var reportBuilder = function reportBuilder(message, isError) {
    isError ? (0, _logger.error)(message) : (0, _logger.warn)(message);

    return { error: isError, message: message };
};

var getMatches = function getMatches(string, regex) {
    var index = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;

    var matches = [];
    var match = void 0;
    while (match = regex.exec(string)) {
        matches.push(match[index]);
    }
    return matches;
};

var noMissingTag = function noMissingTag(string) {
    var openTags = [];
    var closedTags = [];

    _lodash2.default.forEach(getMatches(string, /<([a-zA-Z]+)>/g, 1), function (tag) {
        openTags.push(tag);
    });

    _lodash2.default.forEach(getMatches(string, /<\/([a-zA-Z]+)>/g, 1), function (tag) {
        closedTags.push(tag);
    });

    var tags = _lodash2.default.mergeWith(_lodash2.default.countBy(openTags), _lodash2.default.countBy(closedTags), function (objValue, srcValue) {
        return { open: objValue || 0, closed: srcValue || 0 };
    });

    return _lodash2.default.every(tags, function (tag) {
        return tag.open && tag.closed && tag.open === tag.closed;
    });
};

var rules = {
    htmlValid: {
        test: function test(string) {
            return (0, _isHtml2.default)('<p>' + string + '</p>');
        },
        report: function report(value, key, lang, isError) {
            return reportBuilder(lang + ': translation for key \'' + key + '\' is not html valid \n  ' + value, isError);
        }
    },
    noMissingTag: {
        test: noMissingTag,
        report: function report(value, key, lang, isError) {
            return reportBuilder(lang + ': translation for key \'' + key + '\' has missing html tag \n  ' + value, isError);
        }
    },
    doubleSpace: {
        test: function test(string) {
            return !/(?:\s|\\u00a0){2}/.test(string);
        },
        report: function report(value, key, lang, isError) {
            return reportBuilder(lang + ': translation for key \'' + key + '\' has double white space \n  ' + value, isError);
        }

    },
    htmlDoubleSpace: {
        test: function test(string) {
            return !/(?:\s|\\u00a0)<\/?[a-zA-Z]+>(?:\s|\\u00a0)/.test(string);
        },
        report: function report(value, key, lang, isError) {
            return reportBuilder(lang + ': translation for key \'' + key + '\' has double white space with inner html tag \n  ' + value, isError);
        }
    }
};

exports.default = function (langs) {
    var isError = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

    var reports = [];
    _lodash2.default.forEach(langs, function (lang, langName) {
        return _lodash2.default.forEach((0, _utils.flatten)(lang), function (value, key) {
            return _lodash2.default.forEach(rules, function (_ref) {
                var test = _ref.test,
                    report = _ref.report;

                if (!test(value)) {
                    reports.push(report(value, key, langName, isError));
                }
            });
        });
    });

    return reports;
};