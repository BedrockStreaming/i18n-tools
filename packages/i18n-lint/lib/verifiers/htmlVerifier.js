'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validateHTML = undefined;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _isHtml = require('is-html');

var _isHtml2 = _interopRequireDefault(_isHtml);

var _utils = require('../utils');

var _logger = require('../logger');

var _reportBuilder = require('../reporters/reportBuilder');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getMatches = function getMatches(string, regex) {
  var index = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;

  var matches = [];
  var match = regex.exec(string);
  while (match) {
    matches.push(match[index]);
    match = regex.exec(string);
  }

  return matches;
};

var noMissingTag = function noMissingTag(string) {
  var openTags = [];
  var closedTags = [];

  _lodash2.default.forEach(getMatches(string, /<([a-zA-Z]+)>/g, 1), function (tag) {
    openTags.push(tag);
  });

  _lodash2.default.forEach(getMatches(string, /<\/([a-zA-Z]+)>/g, 1), function (tag) {
    closedTags.push(tag);
  });

  var tags = _lodash2.default.mergeWith(_lodash2.default.countBy(openTags), _lodash2.default.countBy(closedTags), function (objValue, srcValue) {
    return {
      open: objValue || 0,
      closed: srcValue || 0
    };
  });

  return _lodash2.default.every(tags, function (tag) {
    return tag.open && tag.closed && tag.open === tag.closed;
  });
};

var rules = {
  htmlValid: {
    test: function test(string) {
      return (0, _isHtml2.default)('<p>' + string + '</p>');
    },
    message: 'is not html valid'
  },
  noMissingTag: {
    test: noMissingTag,
    message: 'has missing html tag'
  },
  doubleSpace: {
    test: function test(string) {
      return !/(?:\s|\\u00a0){2}/.test(string);
    },
    message: 'has double white space'
  },
  htmlDoubleSpace: {
    test: function test(string) {
      return !/(?:\s|\\u00a0)<\/?[a-zA-Z]+>(?:\s|\\u00a0)/.test(string);
    },
    message: 'has double white space with inner html tag'
  }
};

var validateHTML = exports.validateHTML = function validateHTML(jsonTree, lang) {
  var isError = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var reports = [];

  (0, _logger.info)('Starting lang ' + lang.toUpperCase() + ' \n');
  _lodash2.default.forEach((0, _utils.flatten)(jsonTree), function (value, key) {
    return _lodash2.default.forEach(rules, function (_ref) {
      var test = _ref.test,
          message = _ref.message;

      if (!test(value)) {
        reports.push((0, _reportBuilder.reportBuilder)(lang, key, message, value, isError));
      }
    });
  });

  return reports;
};