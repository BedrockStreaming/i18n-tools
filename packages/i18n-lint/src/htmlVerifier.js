import _ from 'lodash';
import isHtml from 'is-html';
import { flatten } from './utils';
import { error, warn } from './logger';

const reportBuilder = (message, isError) => {
  isError ? error(message) : warn(message);

  return { error: isError, message };
};


const getMatches = (string, regex, index = 1) => {
    const matches = [];
    let match;
    while (match = regex.exec(string)) {
        matches.push(match[index]);
    }
    return matches;
};


const  noMissingTag = string => {
    let openTags = [];
    let closedTags = [];

    _.forEach(getMatches(string, /<([a-zA-Z]+)>/g, 1), tag => {
        openTags.push(tag);
    });

    _.forEach(getMatches(string, /<\/([a-zA-Z]+)>/g, 1), tag => {
        closedTags.push(tag);
    });

    const tags = _.mergeWith(
        _.countBy(openTags),
        _.countBy(closedTags),
        (objValue, srcValue) => ({open: objValue || 0, closed: srcValue || 0})
    );

    return _.every(tags, tag => tag.open && tag.closed && tag.open === tag.closed);
};

const rules = {
  htmlValid: {
    test: string => isHtml(`<p>${string}</p>`),
    report: (value, key, lang, isError) =>
        reportBuilder(
            `${lang}: translation for key '${key}' is not html valid \n  ${value}`,
            isError,
        ),
  },
  noMissingTag: {
    test: noMissingTag,
    report: (value, key, lang, isError) =>
        reportBuilder(
            `${lang}: translation for key '${key}' has missing html tag \n  ${value}`,
            isError,
        ),
  },
  doubleSpace: {
    test: string => !/(?:\s|\\u00a0){2}/.test(string),
    report: (value, key, lang, isError) =>
      reportBuilder(
          `${lang}: translation for key '${key}' has double white space \n  ${value}`,
          isError,
      ),

  },
  htmlDoubleSpace: {
    test: string => !/(?:\s|\\u00a0)<\/?[a-zA-Z]+>(?:\s|\\u00a0)/.test(string),
    report: (value, key, lang, isError) =>
      reportBuilder(
          `${lang}: translation for key '${key}' has double white space with inner html tag \n  ${value}`,
          isError,
      ),
  }
};

export default (langs, isError = true) => {
  const reports = []
  _.forEach(langs, (lang, langName) =>
    _.forEach(flatten(lang), (value, key) =>
      _.forEach(rules, ({ test, report }) => {
        if (!test(value)) {
          reports.push(report(value, key, langName, isError));
        }
      }),
    ),
  );

  return reports;
};
