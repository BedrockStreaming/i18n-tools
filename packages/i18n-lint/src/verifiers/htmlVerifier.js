import _countBy from 'lodash/countBy';
import _every from 'lodash/every';
import _forEach from 'lodash/forEach';
import _mergeWith from 'lodash/mergeWith';
import isHtml from 'is-html';
import { flatten } from '../utils';
import { info } from '../logger';
import { reportBuilder } from '../reporters/reportBuilder';

const getMatches = (string, regex, index = 1) => {
  const matches = [];
  let match = regex.exec(string);
  while (match) {
    matches.push(match[index] ? match[index] : match[match.length - 1]);
    match = regex.exec(string);
  }

  return matches;
};

const noMissingTag = string => {
  const openTags = [];
  const closedTags = [];

  _forEach(getMatches(string, /<(\w+)[^/]*>|<(a)\s+href="\S+"[^/]*>/g, 1), tag => {
    openTags.push(tag);
  });

  _forEach(getMatches(string, /<\/([a-zA-Z]+)>/g, 1), tag => {
    closedTags.push(tag);
  });

  const tags = _mergeWith(_countBy(openTags), _countBy(closedTags), (objValue, srcValue) => ({
    open: objValue || 0,
    closed: srcValue || 0,
  }));

  return _every(tags, tag => tag.open && tag.closed && tag.open === tag.closed);
};

const rules = {
  htmlValid: {
    test: string => isHtml(`<p>${string}</p>`),
    message: 'is not html valid',
  },
  noMissingTag: {
    test: noMissingTag,
    message: 'has missing html tag',
  },
  doubleSpace: {
    test: string => !/(?:\s|\\u00a0){2}/.test(string),
    message: 'has double white space',
  },
  htmlDoubleSpace: {
    test: string => !/(?:\s|\\u00a0)<\/?[a-zA-Z]+>(?:\s|\\u00a0)/.test(string),
    message: 'has double white space with inner html tag',
  },
};

export const validateHTML = (jsonTree, lang, isError = true) => {
  const reports = [];

  info(`Starting lang ${lang.toUpperCase()} \n`);
  _forEach(flatten(jsonTree), (value, key) =>
    _forEach(rules, ({ test, message }) => {
      if (!test(value)) {
        reports.push(reportBuilder(lang, key, message, value, isError));
      }
    }),
  );

  return reports;
};
