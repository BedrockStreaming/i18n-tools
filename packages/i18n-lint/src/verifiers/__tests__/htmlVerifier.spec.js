import { validateHTML } from '../htmlVerifier';

jest.unmock('../htmlVerifier');

const tests = {
  firstTag: 'foo <a> foo',
  lastTag: 'foo </div> foo',
  control: '<div> foo </div>',
  doubleSpace: 'foo  bar',
  doubleSpaceWithNonBreakingSpace: 'foo \u00a0bar',
  doubleSpaceWithHtmlTag: 'foo <foo> bar</foo>',
  doubleSpaceWithHtmlTagClosing: '<foo>foo </foo> bar',
  selfClosing: 'foo <bar/> foo',
  openingWithAttr: 'foo <bar attr="barz"> foo',
};

describe('validateHTML', () => {
  it('should detect html errors', () => {
    expect(validateHTML(tests, 'test', true)).toMatchSnapshot();
  });

  it('should validate anchor with static href value', () => {
    expect(validateHTML({ validAnchor: '<a href="http://localhost">foo</a>' }, 'test', true)).toMatchSnapshot();
  });

  it('should validate anchor with interpolated href value', () => {
    expect(
      validateHTML({ validAnchor: '<a href="%(googlePrivacyUrl)s">Règles de confidentialité</a>' }, 'test', true),
    ).toMatchSnapshot();
  });
});
