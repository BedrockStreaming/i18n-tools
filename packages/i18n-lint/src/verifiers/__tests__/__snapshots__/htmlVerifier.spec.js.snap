// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`htmlVerifier should detect html errors 1`] = `
Array [
  Object {
    "error": true,
    "message": "has missing html tag",
  },
  Object {
    "error": true,
    "message": "has double white space with inner html tag",
  },
  Object {
    "error": true,
    "message": "has missing html tag",
  },
  Object {
    "error": true,
    "message": "has double white space with inner html tag",
  },
  Object {
    "error": true,
    "message": "has double white space",
  },
  Object {
    "error": true,
    "message": "has double white space",
  },
  Object {
    "error": true,
    "message": "has double white space with inner html tag",
  },
  Object {
    "error": true,
    "message": "has double white space with inner html tag",
  },
]
`;
